src/NLog/ILogMessageFormatter.cs

```diff
@@ -0,0 +1,122 @@
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using System;
+using System.Globalization;
+using NLog.StructuredEvents;
+
+namespace NLog
+{
+    interface ILogMessageFormatter
+    {
+        string FormatMessage(LogEventInfo logEvent);
+        ILogMessageFormatter SetFormatProvider(IFormatProvider formatProvider);
+        IFormatProvider FormatProvider { get; }
+        bool HasProperties(LogEventInfo logEvent);
+    }
+
+    class LogMessageFormatter : ILogMessageFormatter
+    {
+        public static readonly LogMessageFormatter Default = new LogMessageFormatter();
+
+        public IFormatProvider FormatProvider { get; private set; }
+
+        public bool HasProperties(LogEventInfo _) { return false; }
+
+        public string FormatMessage(LogEventInfo logEvent)
+        {
+            return string.Format(FormatProvider ?? CultureInfo.CurrentCulture, logEvent.Message, logEvent.Parameters);
+        }
+
+        public ILogMessageFormatter SetFormatProvider(IFormatProvider formatProvider)
+        {
+            if (ReferenceEquals(this, Default) && !ReferenceEquals(formatProvider, null))
+            {
+                return new LogMessageFormatter() { FormatProvider = formatProvider };
+            }
+            else
+            {
+                FormatProvider = formatProvider;
+                return this;
+            }
+        }
+    }
+
+    class LogMessageTemplateFormatter : ILogMessageFormatter
+    {
+        public static readonly LogMessageTemplateFormatter Default = new LogMessageTemplateFormatter();
+
+        public IFormatProvider FormatProvider { get; private set; }
+
+        public bool HasProperties(LogEventInfo logEvent)
+        {
+            return logEvent.Message != null && logEvent.Parameters != null && logEvent.Parameters.Length > 0;
+        }
+
+        public string FormatMessage(LogEventInfo logEvent)
+        {
+            if (HasProperties(logEvent))
+            {
+                var messageTemplate = TemplateParser.TryParseStructuredTemplate(logEvent.Message);
+                if (messageTemplate != null)
+                {
+                    var properties = logEvent.Properties;
+                    for (int i = 0; i < messageTemplate.Holes.Length; ++i)
+                    {
+                        properties[messageTemplate.Holes[i].Name] = logEvent.Parameters[i];
+                    }
+                    return messageTemplate.Render(FormatProvider ?? CultureInfo.CurrentCulture, logEvent.Parameters);
+                }
+                else
+                {
+                    return string.Format(FormatProvider ?? CultureInfo.CurrentCulture, logEvent.Message, logEvent.Parameters);
```

Comment:
Have now changed to re-use the StringBuilder that we have already acquired from the internal StringBuilder-pool.